src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs

```diff
@@ -1427,5 +1432,73 @@ public static void Main()
             var comp = CreateCompilation(src);
             comp.VerifyDiagnostics();
         }
+
+        [Fact]
+        public void WithSemanticModel1()
+        {
+            var src = @"
+data class C(int X, string Y)
+{
+    public static void Main()
+    {
+        var c = new C(0, ""a"");
+        c = c with { X = 2 };
```

Comment:
> with { X = 2 } [](start = 13, length = 15)

Missed the window for this PR, but consider adding tests for other expressions that are used in different initializers but disallowed in `with` initializers. For example `[0] = 2` or values for `IEnumerable/Add` collection.